Observability in microservice based architecture (#546)
microsoft/code-with-engineering-playbook: observability/microservices.md

# Observability in Microservices

> microservices is a very popular software architecture, where the application is arranged as a collection of loosely coupled services. Some of those services can be written in different languages by different teams.

Reviewer: ```suggestion
Microservices is a very popular software architecture, where the application is arranged as a collection of loosely coupled services. Some of those services can be written in different languages by different teams.
```

## Motivations

> We need to consider special cases when creating a microservice architecture from the perspective of observability. The main reason for that is because we want to capture the interaction between those microservices are going to have when making request between each other that can't be correlated.

Reviewer: ```suggestion
We need to consider special cases when creating a microservice architecture from the perspective of observability. We want to capture the interactions when making requests between those microservices and correlate them.
```

Reviewer 2: > 

For technical documentation, a passive voice is generally recommended. I'm guilty of using this tone for documentation, one thing that I'm trying to correct for some time now. One suggested way to reword this sentence - "It is pertinent to consider special cases from the perspective of Observability when creating a microservice architecture."
It's minor, feel free to not use this sentence

> Imagine we have a microservice that accesses the database and retrieve some data as part of a request. This microservice is going to be called by someone else as part of an incoming http request or an internal process being executed. What happen if during the retrieval of the data (or the update of the data) something wrong happen? How we can associate, or correlate, that this particular call failed in the destination microservice?

Reviewer: ```suggestion
Imagine we have a microservice that accesses a database to retrieve some data as part of a request. This microservice is going to be called by someone else as part of an incoming http request or an internal process being executed. What happens if a problem occurs during the retrieval of the data (or the update of the data)? How can we associate, or correlate, that this particular call failed in the destination microservice?
```

> This is a common issue. When calling other microservices, depending on the technology stack we use, we can hide errors and exceptions that might happen on the other side even without wanting it. If we are using a simple REST interface, the other microservice can return to us a 500 HTTP status code and we don't have any idea what happen inside that microservice.

Reviewer: ```suggestion
This is a common issue. When calling other microservices, depending on the technology stack we use, we can accidentally hide errors and exceptions that might happen on the other side. If we are using a simple REST interface, the other microservice can return a 500 HTTP status code and we don't have any idea what happen inside that microservice.
```

> More important, we don't have any way to associate our Correlation Id to whatever happens inside that microservice. Therefore, is so important to have a plan in place to be able to extend your traceability and monitoring efforts, especially when using a microservice architecture.

Reviewer 3: Is it possible to mention in this section about the way(s) to solve the problems you have mentioned above?

Author: One of the things that I mention down there is that most of the framework already support W3C correlation and it's automatically solved.

Reviewer 3: I think what I meant was that there was no transition between the statement of the problem (or motivation to what to solve) and how to solve it, instead it jumped to the implementation on how it works. It is more around the transition, than the content.

Author: Technically the next part is about how to solve that problem and understanding at the same time. I mean, we don't want to have technical implementations here, just explain what and why. What would be your ideal mention here? Talking about how to solve it? libraries?

> More important, we don't have any way to associate our Correlation Id to whatever happens inside that microservice. Therefore, is so important to have a plan in place to be able to extend your traceability and monitoring efforts, especially when using a microservice architecture.

Reviewer 2: Recommendation: I would suggest adding that "the engineer review the APM used to ensure that the context propagation is available out of the box and make sure that no context information is lost between services". I think this is a good callout when the reader starts reading the doc.

Author: That is perfect! Thanks

> ## How to extend your tracing information between microservices

Reviewer 3: Does it make sense to move some of the information here to the "tracing" page?

Author: where is that tracing page? this is related to only where microservices is in place.

Reviewer 3: This is the page: https://github.com/microsoft/code-with-engineering-playbook/blob/master/observability/pillars/tracing.md

Yes, I see your point about microservices. But we don't have a lot of info on that page, so maybe it is an opportunity to start adding information there?

Author: Now that I see the page no. In here we're explaining how to propagate the correlation between microservice, not how to implement tracing. So it's not related with tracing per se.

> The W3C consortium is working on a [Trace Context](https://www.w3.org/TR/trace-context/) definition that we can use when using HTTP as the protocol in our microservice architecture. But let's explain how we can implement this functionality in our software.

Reviewer 2: ```suggestion
The W3C consortium is working on a [Trace Context](https://www.w3.org/TR/trace-context/) definition that can be applied when using HTTP as the protocol in a microservice architecture. But let's explain how we can implement this functionality in our software.
```

> The W3C consortium is working in a [Trace Context](https://www.w3.org/TR/trace-context/) definition that we can use when using HTTP as the protocol in our microservice architecture. But let's explain how we can implement this functionality in our software.

Reviewer: ```suggestion
The W3C consortium is working on a [Trace Context](https://www.w3.org/TR/trace-context/) definition that we can use when using HTTP as the protocol in our microservice architecture. But let's explain how we can implement this functionality in our software.
```

> The main idea behind this is to propagate the correlation information between http request so other pieces of software can read this information and correctly correlate telemetry across microservices.

Reviewer: ```suggestion
The main idea behind this is to propagate the correlation information between HTTP request so other pieces of software can read this information and correctly correlate telemetry across microservices.
```

> The way to propagate this information is to use HTTP Headers for the Correlation Id, parent Correlation Id, etc.

Reviewer 3: I love this information and it is very nicely summarised. Maybe it is worth mentioning that most libraries that support distributed tracing already implement this for you. That way we can keep this practical for people.

Author: Yes, it's mention at the end of the post.

Reviewer 3: Yeah, what I meant was providing a "TL;DR;" earlier and then the extended version in case the person reading would like to understand better how it works.

Author: Well I don't like the TLDR section, it just a page long markdown document, I don't think it make sense, but happy to do what you think is best.

Reviewer 2: I agree with @ivaramme with this one, I;ve added a "Recommendation" snippet above for this

When you are in the scope of a HTTP Request, your tracing system should already have created four properties that you can use to send across your microservices.

- RequestId:0HLQV2BC3VP2T:00000001,
- SpanId:da13aa3c6fd9c146,
- TraceId:f11a03e3f078414fa7c0a0ce568c8b5c,
- ParentId:5076c17d0a604244

This is an example of the four properties you can find which identify the current request.

- RequestId is the unique id that represent the current HTTP Request.
- SpanId is the default automatically generated span. You can have more than one Span that scope different functionality inside your software.
- TraceId represent the id for current log trace.
- ParentId is the parent span id, that in some case can be the same or something different.

## Example

Now we are going to explore an example with 3 microservices that calls to each other in a row.

![image](./microservices.png)

This image is the summary of what is needed in each microservice to propagate the trace-id from A to C.

> The root caller is A and that is why it doesn't have a parent-id, only have a new trace-id. Next, A calls B using HTTP. To propagate the correlation information as part of the request, we are using two new headers, that based on the W3C Correlation specification, are: trace-id and parent-id. So, in this example because A is the root caller, he only sends his own trace-id to the microservice B.

Reviewer: ```suggestion
The root caller is A and that is why it doesn't have a parent-id, only have a new trace-id. Next, A calls B using HTTP. To propagate the correlation information as part of the request, we are using two new headers based on the W3C Correlation specification, trace-id and parent-id. In this example because A is the root caller, A only sends its own trace-id to microservice B.
```

> When the microservice B received the incoming http request, checks the content of these two headers. Reads the content of the trace-id header and sets as his own parent-Id (as shown in the green rectangle inside's B). Also, create a new trace-id to signal that is a new scope for the telemetry. During the execution of the microservice B, it also calls microservice C and it's the same situation. As part of the request include the two headers and propagate trace-id and parent-id as well.

Reviewer: ```suggestion
When microservice B receives the incoming HTTP request, it checks the contents of these two headers. It reads the content of the trace-id header and sets its own parent-id to this trace-id (as shown in the green rectangle inside's B). In addition, it creates a new trace-id to signal that is a new scope for the telemetry. During the execution of microservice B, it also calls microservice C and repeats the pattern. As part of the request it includes the two headers and propagates trace-id and parent-id as well.
```

> Finally, the microservice C, reads the value for the incoming trace-id and sets as his own parent-id, but also create a new trace-id that will use to send telemetry about his own operations.

Reviewer: ```suggestion
Finally, microservice C, reads the value for the incoming trace-id and sets as his own parent-id, but also creates a new trace-id that will use to send telemetry about his own operations.
```

## Summary

> A lot of APM technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SKD and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.

Reviewer 2: ```suggestion
A lot of APM technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SDK and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.
```

> A lot of APM technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SKD and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.

Reviewer 2: Luis, Could you had some notes about what is available out-of-the-box(OOB) for context propagation and what is not. It will be helpful to understand that the context propagation is OOB and valuable to understand what additional information be added (Is it identifying a correlation id and passing it through etc.,). I think it will be valuable to bring your current engagement experience here (generically, without adding the client name)

Author: Thanks @NarmathaBala, I already added a couple of paragraphs with links to the SDKs that already support correlation that can be used in this scenario.

> A lot of APM technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SDK and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.

Reviewer 3: Can you mention what APM stands for, for clarity?

> A lot of Aplication Monitoring (APM) technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SDK and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.

Reviewer 2: ```suggestion
A number of Application Monitoring (APM) technology products already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SDK and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.
```

> A lot of APM technology already support most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). But W3C already proposed recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SKD and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.

Reviewer: ```suggestion
A lot of APM technology already supports most of this Correlation Propagation. The most popular is [OpenZipkin/B3-Propagation](https://github.com/openzipkin/b3-propagation). W3C already proposed a recommendation for the [W3C Trace Context](https://www.w3.org/blog/2019/12/trace-context-enters-proposed-recommendation/), where you can see what SKD and frameworks already support this functionality. It's important to correctly implement the propagation specially when there are different teams that used different technology stacks in the same project.
```
